s3/src/main/scala/quasar/destination/avalanche/s3/AvalancheS3Destination.scala

```diff
@@ -0,0 +1,247 @@
+      : F[Unit] = {
+
+    val dropQuery = dropTableQuery(tableName).update
+
+    val existanceQuery = existanceTableQuery(tableName).query[Int]
+
+    val truncateQuery = truncateTableQuery(tableName).update
+
+    val createQuery = createTableQuery(tableName, cols).update
+
+    val loadQuery = copyQuery(tableName, freshName).update
+
+    for {
+      _ <- debug(s"Table creation query:\n${createQuery.sql}")
+
+      _ <- config.writeMode match {
+        case Replace => debug(s"Drop table query:\n${dropQuery.sql}")
+        case Truncate => debug(s"Truncate table query:\n${truncateQuery.sql}")
+        case Create => ().pure[F]
+      }
+
+      _ <- debugRedacted(s"Load query:\n${loadQuery.sql}")
+
+      load = Sync[ConnectionIO].bracket(createStatement)(st =>
+        for {
+          _ <- Sync[ConnectionIO].delay(st.execute(loadQuery.sql))
+          count <- Sync[ConnectionIO].delay(st.getUpdateCount)
+        } yield count)(st => Sync[ConnectionIO].delay(st.close))
+
+      count <- ((config.writeMode match {
+          case Replace => dropQuery.run *> createQuery.run
+          case Create => createQuery.run
+          case Truncate => (for {
+              exists <- existanceQuery.option
+              result <- if (exists == Some(1)) truncateQuery.run else createQuery.run
+            } yield result)
+        }) *> load).transact(xa)
+
+      _ <- debug(s"Finished load")
+
+      _ <- debug(s"Load result count: $count")
+
+      _ <- debug(s"Finished table copy")
+
+    } yield ()
+  }
+
+  // Ingres accepts double quotes as part of identifiers, but they must
+  // be repeated twice. So we duplicate all quotes
+  // More details:
+  // https://docs.actian.com/avalanche/index.html#page/SQLLanguage%2FRegular_and_Delimited_Identifiers.htm%23ww414482
+  private def escapeIdent(ident: String): String = {
+    val escaped = ident.replace("\"", "\"\"")
+    s""""$escaped""""
+  }
+
+  private def ensureValidTableName(r: ResourcePath): F[TableName] =
+    r match {
+      case file /: ResourcePath.Root => TableName(escapeIdent(file)).pure[F]
+      case _ => MonadResourceErr[F].raiseError(ResourceError.notAResource(r))
+    }
+
+  private def ensureValidColumns(columns: NonEmptyList[Column[ColumnType.Scalar]])
+      : Either[String, NonEmptyList[Fragment]] =
+    columns.traverse(mkColumn(_)).toEither leftMap { errs =>
+      s"Some column types are not supported: ${mkErrorString(errs)}"
+    }
+
+  private def copyQuery(tableName: TableName, fileName: FileName): Fragment = {
+    val table = tableName.value
+    val auth = config.bucketConfig
+    val akey = auth.accessKey
+    val skey = auth.secretKey
+
+    fr"COPY" ++ Fragment.const0(table) ++ fr0"() VWLOAD FROM " ++ s3Path(fileName) ++
+      fr"WITH" ++ pairs(
+      fr"AWS_ACCESS_KEY" -> Fragment.const(s"'${akey.value}'"),
+      fr"AWS_SECRET_KEY" -> Fragment.const(s"'${skey.value}'"),
+      fr"FDELIM" -> fr"','",
+      fr"QUOTE" -> fr"""'"'""") ++ fr", AUTO_DETECT_COMPRESSION"
```

Comment:
Isn't most of this stuff shared completely with the azure destination? Shouldn't it live in core?